File: app/src/main/java/lt/gintaras/tts/engine/Voice.kt
package lt.gintaras.tts.engine

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.Charset

// Bit-for-bit Kotlin port of lt_tts/voice.py
// Decodes gintaras.dta (MFC CMapStringToOb of CMapItem) into a shared pitch-frame pool.

internal object Voice {

    private val TAG_BYTES = byteArrayOf(0x03.toByte(), 0x80.toByte())
    private val SENT = byteArrayOf(0xff.toByte(), 0xff.toByte(), 0xff.toByte(), 0xff.toByte())
    private val ZERO8 = ByteArray(8)

    // ---- frame pool -----------------------------------------------------------------------

    private fun keyBefore(d: ByteArray, i: Int): String? {
        for (L in 1..7) {
            val start = i - 1 - L
            if (start < 0) break
            if (d[start].toInt() and 0xff == L) {
                val slice = d.sliceArray(start + 1 until i)
                if (slice.all { b -> val c = b.toInt() and 0xff; c in 33..126 })
                    return slice.toString(Charsets.ISO_8859_1)
            }
        }
        return null
    }

    fun buildFramePool(d: ByteArray): Map<Int, IntArray> {
        val pool = mutableMapOf<Int, IntArray>()
        val n = d.size
        var i = 0
        while (i < n - 1) {
            // find next 0x03 0x80
            if (d[i] != 0x03.toByte() || d[i + 1] != 0x80.toByte()) { i++; continue }
            val p = i + 2
            if (p + 24 > n) { i++; continue }
            val bb = ByteBuffer.wrap(d, p, 12).order(ByteOrder.LITTLE_ENDIAN)
            val hdrfid = bb.int; val flag = bb.int; val nbytes = bb.int
            if (!(hdrfid in 0..0x1fff && flag in 1..63 && nbytes in 32..0x20000 && nbytes % 2 == 0)) { i++; continue }
            val gap = d.sliceArray(p + 12 until p + 24)
            if (!gap.sliceArray(0..7).contentEquals(ZERO8)) { i++; continue }
            val gapSent = gap.sliceArray(8..11)
            if (!gapSent.contentEquals(SENT) && !gapSent.contentEquals(ByteArray(4))) { i++; continue }
            val key = keyBefore(d, i); if (key == null) { i++; continue }
            if (!key.all { it.isDigit() }) { i++; continue }
            val pcmOff = p + 24
            if (pcmOff + nbytes > n) { i++; continue }
            // mean abs check (reject full-scale garbage)
            val bb2 = ByteBuffer.wrap(d, pcmOff, nbytes).order(ByteOrder.LITTLE_ENDIAN)
            val samps = IntArray(nbytes / 2) { bb2.short.toInt() }
            val meanAbs = samps.sumOf { Math.abs(it).toLong() }.toDouble() / samps.size
            if (meanAbs >= 28000.0) { i++; continue }
            val fid = key.toInt()
            if (fid !in pool) pool[fid] = samps
            i += 2
        }
        return pool
    }

    // ---- units ----------------------------------------------------------------------------

    private fun looksLabel(b: ByteArray): Boolean {
        if (b.isEmpty()) return false
        var hasAlpha = false
        for (byte in b) {
            val c = byte.toInt() and 0xff
            if (c < 0x20) return false
            if (c < 0x80 && !(c.toChar().isLetter() || c.toChar() in "'%-|$^")) return false
            if ((c < 0x80 && c.toChar().isLetter()) || c >= 0x80) hasAlpha = true
        }
        return hasAlpha
    }

    fun parseUnits(d: ByteArray): Map<String, List<Int>> {
        val units = mutableMapOf<String, List<Int>>()
        val n = d.size
        val tagFF = byteArrayOf(0x03, 0x80.toByte(), 0xff.toByte(), 0xff.toByte(), 0xff.toByte(), 0xff.toByte())
        var i = 0
        while (i < n - tagFF.size) {
            // find tag + 0xffffffff
            if (!d.sliceArray(i until i + tagFF.size).contentEquals(tagFF)) { i++; continue }
            val t = i
            var key: String? = null
            for (L in 1..12) {
                val s = t - 1 - L
                if (s < 0) break
                val lb = d[s].toInt() and 0xff
                if (lb == L) {
                    val slice = d.sliceArray(s + 1 until s + 1 + L)
                    if (looksLabel(slice)) {
                        key = slice.toString(Charset.forName("windows-1257")); break
                    }
                }
            }
            if (key == null) { i++; continue }
            val p = i + tagFF.size
            if (p + 20 > n) { i++; continue }
            val bb = ByteBuffer.wrap(d, p, 20).order(ByteOrder.LITTLE_ENDIAN)
            bb.int; bb.int  // i0, i1
            val cnt = bb.int; bb.int; bb.int  // i3, i4
            if (cnt <= 0 || cnt >= 4000) { i++; continue }
            var q = p + 20
            val refs = mutableListOf<Int>()
            var ok = true
            repeat(cnt) {
                if (q >= n || d[q].toInt() and 0xff == 0 || d[q].toInt() and 0xff > 9) { ok = false; return@repeat }
                val L = d[q].toInt() and 0xff
                if (q + 1 + L > n) { ok = false; return@repeat }
                val sid = d.sliceArray(q + 1 until q + 1 + L)
                if (!sid.all { it in 0x30..0x39 }) { ok = false; return@repeat }
                refs.add(sid.toString(Charsets.US_ASCII).toInt())
                q += 1 + L + 8
            }
            if (ok && refs.isNotEmpty()) {
                if (key !in units || refs.size > units[key]!!.size) units[key] = refs
            }
            i++
        }
        return units
    }

    fun parseUnitPads(d: ByteArray): Map<String, Pair<List<Int>, List<Int>>> {
        val out = mutableMapOf<String, Pair<List<Int>, List<Int>>>()
        val n = d.size
        val tagFF = byteArrayOf(0x03, 0x80.toByte(), 0xff.toByte(), 0xff.toByte(), 0xff.toByte(), 0xff.toByte())
        var i = 0
        while (i < n - tagFF.size) {
            if (!d.sliceArray(i until i + tagFF.size).contentEquals(tagFF)) { i++; continue }
            val t = i
            var key: String? = null
            for (L in 1..12) {
                val s = t - 1 - L
                if (s < 0) break
                val lb = d[s].toInt() and 0xff
                if (lb == L) {
                    val slice = d.sliceArray(s + 1 until s + 1 + L)
                    if (looksLabel(slice)) { key = slice.toString(Charset.forName("windows-1257")); break }
                }
            }
            if (key == null) { i++; continue }
            val p = i + tagFF.size
            if (p + 20 > n) { i++; continue }
            val bb = ByteBuffer.wrap(d, p, 20).order(ByteOrder.LITTLE_ENDIAN)
            bb.int; bb.int; val cnt = bb.int; bb.int; bb.int
            if (cnt <= 0 || cnt >= 4000) { i++; continue }
            var q = p + 20
            val pads = mutableListOf<Int>(); val flags = mutableListOf<Int>()
            var ok = true
            repeat(cnt) {
                if (q >= n || d[q].toInt() and 0xff == 0 || d[q].toInt() and 0xff > 9) { ok = false; return@repeat }
                val L = d[q].toInt() and 0xff
                if (q + 1 + L + 8 > n) { ok = false; return@repeat }
                val bb2 = ByteBuffer.wrap(d, q + 1 + L, 8).order(ByteOrder.LITTLE_ENDIAN)
                pads.add(bb2.int); flags.add(bb2.int)
                q += 1 + L + 8
            }
            if (ok && pads.isNotEmpty()) {
                if (key !in out || pads.size > out[key]!!.first.size) out[key] = Pair(pads, flags)
            }
            i++
        }
        return out
    }

    // ---- lazy-loaded voice data -----------------------------------------------------------

    data class VoiceData(val pool: Map<Int, IntArray>, val units: Map<String, List<Int>>)

    @Volatile private var cached: VoiceData? = null

    fun load(): VoiceData {
        cached?.let { return it }
        synchronized(this) {
            cached?.let { return it }
            val d = Assets.bytes("gintaras.dta")
            val pool = buildFramePool(d)
            val units = parseUnits(d)
            return VoiceData(pool, units).also { cached = it }
        }
    }
}

File: app/src/main/python/lt_tts/voice.py
# -*- coding: utf-8 -*-
# Decode wtvlt1.dta = the GINTARAS voice (MFC CMapStringToOb of CMapItem).
#
# Architecture (reverse-engineered 2026-05-31): a SHARED PITCH-FRAME POOL.
#   * A pitch frame is DEFINED once as:  03 80 <i32 fid> <i32 flag> <i32 nbytes> <int16 PCM[nbytes/2]>
#     (fid >= 1).  Frames are shared, so most units only reference them.
#   * A UNIT (syllable/demisyllable, the map key, cp1257) is:
#        <CString key> 03 80  ff ff ff ff  <i32 0> <i32 0>  <i32 framecount>  <i32 0> <i32 0>
#        then framecount frame REFERENCES:  04 "<fid>" <i32 0> <i32 flag>
#     i.e. the unit's audio = concat(pool[fid] for each referenced fid, in order).
#   The unit's CMapItem tag (03 80) is followed by ff ff ff ff (-1); a frame def's
#   03 80 is followed by a positive fid -> that disambiguates the two uses of 03 80.
import re, struct, numpy as np, wave, os
from . import paths

TAG = b'\x03\x80'

def load():
    return open(paths.voice_path(), 'rb').read()

# ---------------------------------------------------------------- frame pool
# A genuine inline frame definition is:
#   03 80 <i32 fid> <i32 flag> <i32 nbytes> <padding> ff ff ff ff <int16 PCM[nbytes/2]>
# The PCM begins right after the ff ff ff ff sentinel (a few bytes of zero padding sit
# between nbytes and the sentinel).  Anchoring on the sentinel + using nbytes as the exact
# length rejects the many spurious 03 80 byte-matches that occur inside PCM data.
SENT = b'\xff\xff\xff\xff'

def _key_before(d, i):
    """MFC CString key serialised right before a frame tag at offset i: BYTE len + ASCII chars.
    Frame keys are short numeric strings ('0'..'6xxx'). Returns the key string or None."""
    for L in range(1, 8):
        if i - 1 - L < 0:
            break
        if d[i - 1 - L] == L and all(33 <= c < 127 for c in d[i - L:i]):
            return d[i - L:i].decode('latin1')
    return None

def build_frame_pool(d):
    """The voice is an MFC CMapStringToOb of (key, frame) pairs. Each frame object:
    03 80 | hdrfid(4) | flag(4) | nbytes(4) | gap(12) | PCM[nbytes]  (PCM at tag+26; gap = 8 zero
    bytes + ff ff ff ff [single pitch period] OR 00 00 00 00 [~100ms segment]).
    CRITICAL: units reference frames by the MAP KEY string that precedes each frame object — NOT
    the header fid (which is an unrelated internal counter). Pool is keyed by int(keystring)."""
    pool = {}
    n = len(d)
    for m in re.finditer(re.escape(TAG), d):
        i = m.start()
        p = m.end()                              # just past 03 80
        if p + 24 > n:
            continue
        hdrfid, flag, nbytes = struct.unpack('<3i', d[p:p+12])
        if not (0 <= hdrfid < 0x2000 and 0 < flag < 64 and 32 <= nbytes <= 0x20000 and nbytes % 2 == 0):
            continue
        gap = d[p+12:p+24]                        # 12-byte gap before PCM (strong validator)
        if gap[:8] != b'\x00' * 8 or gap[8:12] not in (SENT, b'\x00\x00\x00\x00'):
            continue
        key = _key_before(d, i)
        if key is None or not key.isdigit():
            continue
        raw = d[p+24:p+24 + nbytes]
        if len(raw) < nbytes:
            continue
        a = np.frombuffer(raw, dtype='<i2')
        if float(np.mean(np.abs(a))) >= 28000:   # reject only near-full-scale garbage
            continue
        fid = int(key)
        if fid not in pool:
            pool[fid] = a
    return pool

# ---------------------------------------------------------------- units
def looks_label(b):
    if not b:
        return False
    for c in b:
        if c < 0x20:
            return False
        if c < 0x80 and not (chr(c).isalpha() or chr(c) in "'%-|$.^"):
            return False
    return any((c < 0x80 and chr(c).isalpha()) or c >= 0x80 for c in b)

UNIT_HDR = re.compile(re.escape(TAG) + rb'\xff\xff\xff\xff')  # tag + (-1)

def parse_units(d):
    """key -> [fid,...] reference list, in order."""
    units = {}
    for m in UNIT_HDR.finditer(d):
        t = m.start()
        # key is the CString immediately before the tag
        key = None
        for L in range(1, 13):
            s = t - 1 - L
            if s < 0:
                break
            if d[s] == L and looks_label(d[s+1:s+1+L]):
                key = d[s+1:s+1+L].decode('cp1257', 'replace')
                break
        if key is None:
            continue
        p = m.end()                              # after ff ff ff ff
        i0, i1, cnt, i3, i4 = struct.unpack('<5i', d[p:p+20])
        p += 20
        if not (0 < cnt < 4000):
            continue
        refs = []
        ok = True
        for _ in range(cnt):
            if p >= len(d) or d[p] == 0 or d[p] > 9:
                ok = False; break
            L = d[p]; sid = d[p+1:p+1+L]
            if not sid.isdigit():
                ok = False; break
            refs.append(int(sid))
            p += 1 + L + 8                        # id string + i32 pad + i32 flag
        if ok and refs:
            if key not in units or len(refs) > len(units[key]):
                units[key] = refs
    return units

def parse_unit_pads(d):
    """key -> ([pad,...], [flag,...]) per frame reference. ADDITIVE companion to parse_units (does NOT alter
    it): each reference is `<len> "<fid>" <i32 pad> <i32 flag>`. The PAD field stores the engine's per-grain
    a5 (extra-epoch / 'double budget' count) DIRECTLY for the 12 'irregular' demisyllables whose grains carry
    the 0x40 flag bit (the long-vowel nuclei i|/ąą/juo|/u|j/gą|/-dą/-ną/lu|-- + word-final -ka/-rai). For all
    other ~609 units pad is 0 (a5 there is the regular stress-driven rule). FLAG bits: 0x01 voiced, 0x40
    a5-stored-in-pad, 0x80 last-grain-of-unit. RE'd 2026-06-10 (cont.39); see plan-frontend-wiring memory."""
    out = {}
    for m in UNIT_HDR.finditer(d):
        t = m.start()
        key = None
        for L in range(1, 13):
            s = t - 1 - L
            if s < 0:
                break
            if d[s] == L and looks_label(d[s+1:s+1+L]):
                key = d[s+1:s+1+L].decode('cp1257', 'replace')
                break
        if key is None:
            continue
        p = m.end()
        _i0, _i1, cnt, _i3, _i4 = struct.unpack('<5i', d[p:p+20])
        p += 20
        if not (0 < cnt < 4000):
            continue
        pads, flags = [], []
        ok = True
        for _ in range(cnt):
            if p >= len(d) or d[p] == 0 or d[p] > 9:
                ok = False; break
            L = d[p]; sid = d[p+1:p+1+L]
            if not sid.isdigit():
                ok = False; break
            pad, flag = struct.unpack('<2i', d[p+1+L:p+1+L+8])
            pads.append(pad); flags.append(flag)
            p += 1 + L + 8
        if ok and pads and (key not in out or len(pads) > len(out[key][0])):
            out[key] = (pads, flags)
    return out


def synth_unit(pool, refs):
    segs = [pool[f] for f in refs if f in pool]
    if not segs:
        return np.array([], dtype='<i2'), 0
    return np.concatenate(segs), sum(1 for f in refs if f in pool)

def save_wav(path, sig, sr=22050):
    wv = wave.open(path, 'wb'); wv.setnchannels(1); wv.setsampwidth(2); wv.setframerate(sr)
    wv.writeframes(np.asarray(sig, dtype='<i2').tobytes()); wv.close()
